src/compiler/checker.ts

```diff
@@ -34305,16 +34305,51 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return false;
     }
 
-    function checkAssertionWorker(errNode: Node, type: TypeNode, expression: UnaryExpression | Expression, checkMode?: CheckMode) {
-        let exprType = checkExpression(expression, checkMode);
+    function checkAssertionWorker(node: JSDocTypeAssertion | AssertionExpression, checkMode: CheckMode | undefined) {
+        let type: TypeNode;
+        let expression: Expression;
+        switch (node.kind) {
+            case SyntaxKind.AsExpression:
+            case SyntaxKind.TypeAssertionExpression:
+                type = node.type;
+                expression = node.expression;
+                break;
+            case SyntaxKind.ParenthesizedExpression:
+                type = getJSDocTypeAssertionType(node);
+                expression = node.expression;
+                break;

```

Comment:
Maybe extract this into a helper?
```ts
function getTypeAndExpressionFromAssertion(node: AssertionExpression | JSDocTypeAssertion): { type: TypeNode, expression: Expression } {
        switch (node.kind) {
            case SyntaxKind.AsExpression:
            case SyntaxKind.TypeAssertionExpression:
                return node;
            case SyntaxKind.ParenthesizedExpression:
                return { type: getJSDocTypeAssertionType(node), expression: node.expression };
        }
}
```

So this and the other call site can just be
```ts
const {type, expression} = getTypeAndExpressionFromAssertion(node);
```